File: src/系统/08．任务系统/04．击杀任务进度.lua
local ____lualib = require("lualib_bundle")
local __TS__StringSplit = ____lualib.__TS__StringSplit
local __TS__StringTrim = ____lualib.__TS__StringTrim
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local ____exports = {}
local ____02_FF0E_4EFB_52A1_914D_7F6E_8868 = require("系统.08．任务系统.00．配置表.02．任务配置表")
local _____4EFB_52A1_914D_7F6E_5217_8868 = ____02_FF0E_4EFB_52A1_914D_7F6E_8868["任务配置列表"]
local ____01_FF0E_4EFB_52A1_6570_636E = require("系统.08．任务系统.01．任务数据")
local questDB = ____01_FF0E_4EFB_52A1_6570_636E.questDB
local QuestStatus = ____01_FF0E_4EFB_52A1_6570_636E.QuestStatus
local ____02_FF0E_4EFB_52A1_7BA1_7406_5668 = require("系统.08．任务系统.02．任务管理器")
local _____89E6_53D1_4EFB_52A1UI_5237_65B0 = ____02_FF0E_4EFB_52A1_7BA1_7406_5668["触发任务UI刷新"]
---
-- @noSelfInFile
local jass = require("jass.common")
local GetOwningPlayer = jass.GetOwningPlayer
local GetPlayerController = jass.GetPlayerController
local GetPlayerId = jass.GetPlayerId
local GetUnitTypeId = jass.GetUnitTypeId
local Player = jass.Player
local MAP_CONTROL_USER = jass.MAP_CONTROL_USER
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_0.registerDeathListener
local ____require_result_1 = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版")
local fourCCToStringSafe = ____require_result_1.fourCCToStringSafe
local stringToFourCCSafe = ____require_result_1.stringToFourCCSafe
local ____require_result_2 = require("lib.扩展函数.物品相关函数.物品判断函数")
local UnitHasItemOfTypeBJ = ____require_result_2.UnitHasItemOfTypeBJ
local ____require_result_3 = require("系统.00．核心系统.00．玩家系统.00．英雄注册联动.00．玩家英雄获取桥接")
local getRegisteredPlayerHero = ____require_result_3.getRegisteredPlayerHero
local ____require_result_4 = require("系统.09．表现系统.06．广播提示消息.index")
local _____53D1_9001_5934_50CF_63D0_793A_7ED9_73A9_5BB6 = ____require_result_4["发送头像提示给玩家"]
local ____require_result_5 = require("系统.09．表现系统.06．广播提示消息.00．常量定义")
local _____5E7F_64AD_63D0_793A_73A9_5BB6_69FD_6570 = ____require_result_5["广播提示玩家槽数"]
local _____5E7F_64AD_63D0_793A_5587_53ED_5934_50CF = ____require_result_5["广播提示喇叭头像"]
local _____5DF2_521D_59CB_5316 = false
local function _____662F_51FB_6740_4EFB_52A1_914D_7F6E(_____914D_7F6E)
    return (_____914D_7F6E["类型"] == "击杀" or _____914D_7F6E["类型"] == "目标击杀") and _____914D_7F6E["目标单位"] ~= nil and _____914D_7F6E["目标单位"] ~= ""
end
local function _____76EE_6807_5355_4F4D_5339_914D(_____76EE_6807_5355_4F4D_5217_8868, _____6B7B_4EA1_5355_4F4D_4EE3_7801)
    local _____5355_4F4D_4EE3_7801_5217_8868 = __TS__StringSplit(_____76EE_6807_5355_4F4D_5217_8868, "|")
    do
        local i = 0
        while i < #_____5355_4F4D_4EE3_7801_5217_8868 do
            if __TS__StringTrim(_____5355_4F4D_4EE3_7801_5217_8868[i + 1]) == _____6B7B_4EA1_5355_4F4D_4EE3_7801 then
                return true
            end
            i = i + 1
        end
    end
    return false
end
local function _____73A9_5BB6_82F1_96C4_6EE1_8DB3_51FB_6740_643A_5E26_6761_4EF6(_____73A9_5BB6, _____643A_5E26_7269_54C1_5217_8868)
    if not _____643A_5E26_7269_54C1_5217_8868 or _____643A_5E26_7269_54C1_5217_8868 == "" then
        return true
    end
    local _____82F1_96C4 = getRegisteredPlayerHero(_____73A9_5BB6)
    if _____82F1_96C4 == nil or _____82F1_96C4 == 0 then
        return false
    end
    local _____7269_54C1_4EE3_7801_5217_8868 = __TS__StringSplit(_____643A_5E26_7269_54C1_5217_8868, "|")
    do
        local i = 0
        while i < #_____7269_54C1_4EE3_7801_5217_8868 do
            local _____7269_54C1_7C7B_578BID = stringToFourCCSafe(__TS__StringTrim(_____7269_54C1_4EE3_7801_5217_8868[i + 1]))
            if _____7269_54C1_7C7B_578BID ~= 0 and UnitHasItemOfTypeBJ(_____82F1_96C4, _____7269_54C1_7C7B_578BID) then
                return true
            end
            i = i + 1
        end
    end
    return false
end
local function _____67E5_627E_4EFB_52A1_914D_7F6E(_____4EFB_52A1ID)
    do
        local i = 0
        while i < #_____4EFB_52A1_914D_7F6E_5217_8868 do
            local _____914D_7F6E = _____4EFB_52A1_914D_7F6E_5217_8868[i + 1]
            if _____914D_7F6E["任务ID"] ~= nil and tostring(_____914D_7F6E["任务ID"]) == _____4EFB_52A1ID then
                return _____914D_7F6E
            end
            i = i + 1
        end
    end
    return nil
end
local function _____5956_52B1_9762_5411_6240_6709_73A9_5BB6(_____5956_52B1)
    if not _____5956_52B1 or _____5956_52B1 == "" then
        return true
    end
    if (string.find(_____5956_52B1, "所有玩家", nil, true) or 0) - 1 >= 0 or (string.find(_____5956_52B1, "all", nil, true) or 0) - 1 >= 0 then
        return true
    end
    return (string.find(_____5956_52B1, "完成任务的玩家", nil, true) or 0) - 1 < 0 and (string.find(_____5956_52B1, "Player", nil, true) or 0) - 1 < 0
end
local function _____6784_5EFA_51FB_6740_8FDB_5EA6_64AD_62A5(_____914D_7F6E, _____5F53_524D_6570_91CF, _____9700_6C42_6570_91CF)
    local _____8FDB_5EA6_6A21_677F = _____914D_7F6E["进度文本"] or ((_____914D_7F6E["名称"] or "击杀目标") .. "N/") .. tostring(_____9700_6C42_6570_91CF)
    local _____6807_8BB0_4F4D_7F6E = (string.find(_____8FDB_5EA6_6A21_677F, "N/", nil, true) or 0) - 1
    if _____6807_8BB0_4F4D_7F6E >= 0 then
        return (("|cffffff00『任务进度』：|r" .. __TS__StringSubstring(_____8FDB_5EA6_6A21_677F, 0, _____6807_8BB0_4F4D_7F6E)) .. ("|cffffcc00" .. tostring(_____5F53_524D_6570_91CF)) .. "|r/") .. __TS__StringSubstring(_____8FDB_5EA6_6A21_677F, _____6807_8BB0_4F4D_7F6E + 2)
    end
    return ((((("|cffffff00『任务进度』：|r" .. _____8FDB_5EA6_6A21_677F) .. " |cffffcc00") .. tostring(_____5F53_524D_6570_91CF)) .. "/") .. tostring(_____9700_6C42_6570_91CF)) .. "|r"
end
local function _____64AD_62A5_51FB_6740_4EFB_52A1_8FDB_5EA6(_____73A9_5BB6ID, _____914D_7F6E, _____5F53_524D_6570_91CF, _____9700_6C42_6570_91CF)
    if _____9700_6C42_6570_91CF <= 1 then
        return
    end
    local _____6587_672C = _____6784_5EFA_51FB_6740_8FDB_5EA6_64AD_62A5(_____914D_7F6E, _____5F53_524D_6570_91CF, _____9700_6C42_6570_91CF)
    if not _____5956_52B1_9762_5411_6240_6709_73A9_5BB6(_____914D_7F6E["奖励"]) then
        _____53D1_9001_5934_50CF_63D0_793A_7ED9_73A9_5BB6(
            Player(_____73A9_5BB6ID),
            _____5E7F_64AD_63D0_793A_5587_53ED_5934_50CF,
            _____6587_672C
        )
        return
    end
    do
        local _____76EE_6807_73A9_5BB6ID = 0
        while _____76EE_6807_73A9_5BB6ID < _____5E7F_64AD_63D0_793A_73A9_5BB6_69FD_6570 do
            _____53D1_9001_5934_50CF_63D0_793A_7ED9_73A9_5BB6(
                Player(_____76EE_6807_73A9_5BB6ID),
                _____5E7F_64AD_63D0_793A_5587_53ED_5934_50CF,
                _____6587_672C
            )
            _____76EE_6807_73A9_5BB6ID = _____76EE_6807_73A9_5BB6ID + 1
        end
    end
end
local function _____589E_52A0_51FB_6740_4EFB_52A1_8FDB_5EA6(_____73A9_5BB6ID, _____4EFB_52A1, _____914D_7F6E, _____6B7B_4EA1_5355_4F4D_4EE3_7801)
    if not _____914D_7F6E["目标单位"] or not _____76EE_6807_5355_4F4D_5339_914D(_____914D_7F6E["目标单位"], _____6B7B_4EA1_5355_4F4D_4EE3_7801) then
        return
    end
    local _____76EE_6807 = _____4EFB_52A1.objectives[1]
    if _____914D_7F6E["目标单位分别击杀"] == true then
        _____76EE_6807 = nil
        local _____76EE_6807ID = "kill_" .. _____6B7B_4EA1_5355_4F4D_4EE3_7801
        do
            local i = 0
            while i < #_____4EFB_52A1.objectives do
                if _____4EFB_52A1.objectives[i + 1] ~= nil and _____4EFB_52A1.objectives[i + 1].id == _____76EE_6807ID then
                    _____76EE_6807 = _____4EFB_52A1.objectives[i + 1]
                    break
                end
                i = i + 1
            end
        end
    end
    if not _____76EE_6807 or _____76EE_6807.current >= _____76EE_6807.required then
        return
    end
    local _____65B0_8FDB_5EA6 = _____76EE_6807.current + 1
    if questDB:updateObjective(_____73A9_5BB6ID, _____4EFB_52A1.id, _____76EE_6807.id, _____65B0_8FDB_5EA6) then
        _____89E6_53D1_4EFB_52A1UI_5237_65B0(_____73A9_5BB6ID, _____4EFB_52A1.id)
        local _____603B_8FDB_5EA6 = 0
        local _____603B_9700_6C42 = 0
        do
            local i = 0
            while i < #_____4EFB_52A1.objectives do
                do
                    local _____5F53_524D_76EE_6807 = _____4EFB_52A1.objectives[i + 1]
                    if not _____5F53_524D_76EE_6807 then
                        goto __continue36
                    end
                    _____603B_8FDB_5EA6 = _____603B_8FDB_5EA6 + (_____5F53_524D_76EE_6807.id == _____76EE_6807.id and _____65B0_8FDB_5EA6 or _____5F53_524D_76EE_6807.current)
                    _____603B_9700_6C42 = _____603B_9700_6C42 + _____5F53_524D_76EE_6807.required
                end
                ::__continue36::
                i = i + 1
            end
        end
        _____64AD_62A5_51FB_6740_4EFB_52A1_8FDB_5EA6(_____73A9_5BB6ID, _____914D_7F6E, _____603B_8FDB_5EA6, _____603B_9700_6C42)
    end
end
local function ____on_4EFB_52A1_76EE_6807_5355_4F4D_6B7B_4EA1(_____6B7B_4EA1_5355_4F4D, _____51FB_6740_5355_4F4D)
    if not _____6B7B_4EA1_5355_4F4D or not _____51FB_6740_5355_4F4D then
        return
    end
    local _____51FB_6740_73A9_5BB6 = GetOwningPlayer(_____51FB_6740_5355_4F4D)
    if not _____51FB_6740_73A9_5BB6 or GetPlayerController(_____51FB_6740_73A9_5BB6) ~= MAP_CONTROL_USER then
        return
    end
    local _____73A9_5BB6ID = GetPlayerId(_____51FB_6740_73A9_5BB6)
    local _____6B7B_4EA1_5355_4F4D_4EE3_7801 = fourCCToStringSafe(GetUnitTypeId(_____6B7B_4EA1_5355_4F4D))
    local _____8FDB_884C_4E2D_4EFB_52A1 = questDB:getPlayerActiveQuests(_____73A9_5BB6ID)
    do
        local i = 0
        while i < #_____8FDB_884C_4E2D_4EFB_52A1 do
            do
                local _____4EFB_52A1 = _____8FDB_884C_4E2D_4EFB_52A1[i + 1]
                if not _____4EFB_52A1 or _____4EFB_52A1.status ~= QuestStatus.IN_PROGRESS then
                    goto __continue42
                end
                local _____914D_7F6E = _____67E5_627E_4EFB_52A1_914D_7F6E(_____4EFB_52A1.id)
                if not _____914D_7F6E or not _____662F_51FB_6740_4EFB_52A1_914D_7F6E(_____914D_7F6E) then
                    goto __continue42
                end
                if not _____73A9_5BB6_82F1_96C4_6EE1_8DB3_51FB_6740_643A_5E26_6761_4EF6(_____51FB_6740_73A9_5BB6, _____914D_7F6E["击杀携带物品"]) then
                    goto __continue42
                end
                _____589E_52A0_51FB_6740_4EFB_52A1_8FDB_5EA6(_____73A9_5BB6ID, _____4EFB_52A1, _____914D_7F6E, _____6B7B_4EA1_5355_4F4D_4EE3_7801)
            end
            ::__continue42::
            i = i + 1
        end
    end
end
____exports["初始化击杀任务进度"] = function()
    if _____5DF2_521D_59CB_5316 then
        return
    end
    _____5DF2_521D_59CB_5316 = true
    registerDeathListener(____on_4EFB_52A1_76EE_6807_5355_4F4D_6B7B_4EA1)
end
return ____exports

File: TS/系统/08．任务系统/04．击杀任务进度.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const GetOwningPlayer = jass.GetOwningPlayer as (unit: any) => any;
const GetPlayerController = jass.GetPlayerController as (player: any) => any;
const GetPlayerId = jass.GetPlayerId as (player: any) => number;
const GetUnitTypeId = jass.GetUnitTypeId as (unit: any) => number;
const Player = jass.Player as (playerId: number) => any;
const MAP_CONTROL_USER = jass.MAP_CONTROL_USER;

const { registerDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};
const { fourCCToStringSafe, stringToFourCCSafe } = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版") as {
  fourCCToStringSafe: (this: void, fourcc: number) => string;
  stringToFourCCSafe: (this: void, rawId: string | undefined | null) => number;
};
const { UnitHasItemOfTypeBJ } = require("lib.扩展函数.物品相关函数.物品判断函数") as {
  UnitHasItemOfTypeBJ: (this: void, unit: any, itemTypeId: number) => boolean;
};
const { getRegisteredPlayerHero } = require("系统.00．核心系统.00．玩家系统.00．英雄注册联动.00．玩家英雄获取桥接") as {
  getRegisteredPlayerHero: (this: void, player: any) => any;
};
const { 发送头像提示给玩家 } = require("系统.09．表现系统.06．广播提示消息.index") as {
  发送头像提示给玩家: (this: void, 目标玩家: any, 头像路径: string, 文本: string, 持续时间?: number) => void;
};
const { 广播提示玩家槽数, 广播提示喇叭头像 } = require("系统.09．表现系统.06．广播提示消息.00．常量定义") as {
  广播提示玩家槽数: number;
  广播提示喇叭头像: string;
};

import { 任务配置, 任务配置列表 } from "./00．配置表/02．任务配置表";
import { questDB, QuestData, QuestStatus } from "./01．任务数据";
import { 触发任务UI刷新 } from "./02．任务管理器";

let 已初始化 = false;

function 是击杀任务配置(this: void, 配置: 任务配置): boolean {
  return (配置.类型 === "击杀" || 配置.类型 === "目标击杀")
    && 配置.目标单位 != null
    && 配置.目标单位 !== "";
}

function 目标单位匹配(this: void, 目标单位列表: string, 死亡单位代码: string): boolean {
  const 单位代码列表 = 目标单位列表.split("|");
  for (let i = 0; i < 单位代码列表.length; i++) {
    if (单位代码列表[i].trim() === 死亡单位代码) return true;
  }
  return false;
}

function 玩家英雄满足击杀携带条件(this: void, 玩家: any, 携带物品列表: string | undefined): boolean {
  if (!携带物品列表 || 携带物品列表 === "") return true;
  const 英雄 = getRegisteredPlayerHero(玩家);
  if (英雄 == null || 英雄 === 0) return false;

  const 物品代码列表 = 携带物品列表.split("|");
  for (let i = 0; i < 物品代码列表.length; i++) {
    const 物品类型ID = stringToFourCCSafe(物品代码列表[i].trim());
    if (物品类型ID !== 0 && UnitHasItemOfTypeBJ(英雄, 物品类型ID)) return true;
  }
  return false;
}

function 查找任务配置(this: void, 任务ID: string): 任务配置 | null {
  for (let i = 0; i < 任务配置列表.length; i++) {
    const 配置 = 任务配置列表[i];
    if (配置.任务ID != null && 配置.任务ID.toString() === 任务ID) return 配置;
  }
  return null;
}

function 奖励面向所有玩家(this: void, 奖励: string | undefined): boolean {
  if (!奖励 || 奖励 === "") return true;
  if (奖励.indexOf("所有玩家") >= 0 || 奖励.indexOf("all") >= 0) return true;
  return 奖励.indexOf("完成任务的玩家") < 0 && 奖励.indexOf("Player") < 0;
}

function 构建击杀进度播报(this: void, 配置: 任务配置, 当前数量: number, 需求数量: number): string {
  const 进度模板 = 配置.进度文本 || `${配置.名称 || "击杀目标"}N/${需求数量}`;
  const 标记位置 = 进度模板.indexOf("N/");
  if (标记位置 >= 0) {
    return "|cffffff00『任务进度』：|r"
      + 进度模板.substring(0, 标记位置)
      + `|cffffcc00${当前数量}|r/`
      + 进度模板.substring(标记位置 + 2);
  }
  return `|cffffff00『任务进度』：|r${进度模板} |cffffcc00${当前数量}/${需求数量}|r`;
}

function 播报击杀任务进度(this: void, 玩家ID: number, 配置: 任务配置, 当前数量: number, 需求数量: number): void {
  if (需求数量 <= 1) return;
  const 文本 = 构建击杀进度播报(配置, 当前数量, 需求数量);
  if (!奖励面向所有玩家(配置.奖励)) {
    发送头像提示给玩家(Player(玩家ID), 广播提示喇叭头像, 文本);
    return;
  }
  for (let 目标玩家ID = 0; 目标玩家ID < 广播提示玩家槽数; 目标玩家ID++) {
    发送头像提示给玩家(Player(目标玩家ID), 广播提示喇叭头像, 文本);
  }
}

function 增加击杀任务进度(this: void, 玩家ID: number, 任务: QuestData, 配置: 任务配置, 死亡单位代码: string): void {
  if (!配置.目标单位 || !目标单位匹配(配置.目标单位, 死亡单位代码)) return;
  let 目标 = 任务.objectives[0];
  if (配置.目标单位分别击杀 === true) {
    目标 = null as any;
    const 目标ID = "kill_" + 死亡单位代码;
    for (let i = 0; i < 任务.objectives.length; i++) {
      if (任务.objectives[i] != null && 任务.objectives[i].id === 目标ID) {
        目标 = 任务.objectives[i];
        break;
      }
    }
  }
  if (!目标 || 目标.current >= 目标.required) return;
  const 新进度 = 目标.current + 1;
  if (questDB.updateObjective(玩家ID, 任务.id, 目标.id, 新进度)) {
    触发任务UI刷新(玩家ID, 任务.id);
    let 总进度 = 0;
    let 总需求 = 0;
    for (let i = 0; i < 任务.objectives.length; i++) {
      const 当前目标 = 任务.objectives[i];
      if (!当前目标) continue;
      总进度 += 当前目标.id === 目标.id ? 新进度 : 当前目标.current;
      总需求 += 当前目标.required;
    }
    播报击杀任务进度(玩家ID, 配置, 总进度, 总需求);
  }
}

function on任务目标单位死亡(this: void, 死亡单位: any, 击杀单位: any): void {
  if (!死亡单位 || !击杀单位) return;
  const 击杀玩家 = GetOwningPlayer(击杀单位);
  if (!击杀玩家 || GetPlayerController(击杀玩家) !== MAP_CONTROL_USER) return;
  const 玩家ID = GetPlayerId(击杀玩家);
  const 死亡单位代码 = fourCCToStringSafe(GetUnitTypeId(死亡单位));
  const 进行中任务 = questDB.getPlayerActiveQuests(玩家ID);
  for (let i = 0; i < 进行中任务.length; i++) {
    const 任务 = 进行中任务[i];
    if (!任务 || 任务.status !== QuestStatus.IN_PROGRESS) continue;
    const 配置 = 查找任务配置(任务.id);
    if (!配置 || !是击杀任务配置(配置)) continue;
    if (!玩家英雄满足击杀携带条件(击杀玩家, 配置.击杀携带物品)) continue;
    增加击杀任务进度(玩家ID, 任务, 配置, 死亡单位代码);
  }
}

export function 初始化击杀任务进度(this: void): void {
  if (已初始化) return;
  已初始化 = true;
  registerDeathListener(on任务目标单位死亡);
}

export {};
